File: scripts_c/ga.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>
#include <string.h>

// Variaveis globais para gerarGrid

float mutation_rate = 0.05;
int pop_size = 30;
int iterations = 100000;
int n = 100;
float target_distance = 1800;

// Estrutura para representar um ponto 3D
typedef struct
{
    int id;
    double x;
    double y;
    double z;
} Point;

// Estrutura para representar um indivíduo (caminho)
typedef struct
{
    int path[100];
    double fitness;
} Individual;

// Calcula a distância euclidiana entre dois pontos 3D
double distance(Point p1, Point p2)
{
    float distancia;
    distancia = sqrt(pow(p2.x - p1.x, 2) + pow(p2.y - p1.y, 2) + pow(p2.z - p1.z, 2));
    return distancia;
}

// Calcula o comprimento do caminho de um indivíduo
double calculate_fitness(Individual individual, Point points[])
{
    double length = 0.0;
    for (int i = 0; i < n - 1; i++)
    {
        length += distance(points[individual.path[i]], points[individual.path[i + 1]]);
    }
    length += distance(points[individual.path[n - 1]], points[individual.path[0]]); // fecha o ciclo
    return length;
}

// Inicializa uma população de indivíduos com caminhos aleatórios
void initialize_population(Individual population[], Point points[])
{
    // Loop através de cada indivíduo na população
    for (int i = 0; i < pop_size; i++)
    {
        // Inicializar caminho com ordem sequencial
        for (int j = 0; j < n; j++)
        {
            population[i].path[j] = j;
        }

        // Embaralhar os pontos exceto o primeiro (ponto de partida)
        for (int j = n - 1; j > 0; j--)
        {
            // Gerar um índice aleatório dentro do intervalo [0, j]
            int k = rand() % (j + 1);

            // Trocar os valores dos índices j e k (exceto para o ponto inicial)
            if (k != 0) // Não embaralhar o ponto de partida
            {
                int temp = population[i].path[j];
                population[i].path[j] = population[i].path[k];
                population[i].path[k] = temp;
            }
        }

        // Calcular o fitness do indivíduo
        population[i].fitness = calculate_fitness(population[i], points);
    }
}

// Seleciona indivíduos para reprodução usando torneio binário
Individual tournament_selection(Individual population[])
{
    Individual parent1 = population[rand() % pop_size];
    Individual parent2 = population[rand() % pop_size];
    return (parent1.fitness < parent2.fitness) ? parent1 : parent2;
}

// Realiza o cruzamento entre dois indivíduos para gerar um novo indivíduo
Individual crossover(Individual parent1, Individual parent2, Point *points)
{
    int start = rand() % n;
    int end = rand() % n;
    if (start > end)
    {
        int temp = start;
        start = end;
        end = temp;
    }
    Individual child;
    for (int i = 0; i < n; i++)
    {
        if (i >= start && i <= end)
        {
            child.path[i] = parent1.path[i];
        }
        else
        {
            child.path[i] = -1;
        }
    }
    int index = 0;
    for (int i = 0; i < n; i++)
    {
        if (index == start)
        {
            index = end + 1;
        }
        int gene = parent2.path[i];
        if (child.path[index] == -1)
        {
            bool contains = false;
            for (int j = 0; j < n; j++)
            {
                if (child.path[j] == gene)
                {
                    contains = true;
                    break;
                }
            }
            if (!contains)
            {
                child.path[index++] = gene;
            }
        }
    }
    child.fitness = calculate_fitness(child, points);
    return child;
}

// Realiza mutação por deslocamento simples em um indivíduo
void mutate_deslocamento_simples(Individual *individual)
{
    if ((double)rand() / RAND_MAX < mutation_rate)
    {
        int start = rand() % (n - 1) + 1;  // Ignorar o ponto de partida
        int length = rand() % (n - start); // Comprimento do deslocamento

        int temp[n];
        memcpy(temp, individual->path, sizeof(temp)); // Copia o caminho atual

        // Desloca a sub-rota em 'length' posições
        for (int i = 0; i < length; i++)
            individual->path[start + i] = temp[start + length - i - 1];
    }
}

void mutate_plus(Individual *individual)
{
    if ((double)rand() / RAND_MAX < 0.1)
    {
        int num_reversals = rand() % 3 + 1; // Escolha aleatoriamente entre 1, 2 ou 3 índices para inverter
        for (int i = 0; i < num_reversals; i++)
        {
            int start = rand() % (n - 2) + 1; // Ignorar o ponto de partida e o último ponto
            int end = start + 1;
            int temp = individual->path[start];
            individual->path[start] = individual->path[end];
            individual->path[end] = temp;
        }
    }
}

// Realiza mutação em um indivíduo
void mutate(Individual *individual)
{
    if ((double)rand() / RAND_MAX < mutation_rate)
    {
        int start = rand() % (n - 1) + 1; // Ignorar o ponto de partida
        int end = rand() % (n - 1) + 1;   // Ignorar o ponto de partida
        if (start > end)
        {
            int temp = start;
            start = end;
            end = temp;
        }
        while (start < end)
        {
            int temp = individual->path[start];
            individual->path[start] = individual->path[end];
            individual->path[end] = temp;
            start++;
            end--;
        }
    }
}

// Evolui a população por uma geração e verifica o critério de parada
bool evolve_population(Individual population[], Point points[])
{
    Individual new_population[pop_size];
    for (int i = 0; i < pop_size; i++)
    {
        Individual parent1 = tournament_selection(population);
        Individual parent2 = tournament_selection(population);
        Individual child = crossover(parent1, parent2, points);
        mutate(&child);
        new_population[i] = child;
        if (child.fitness < target_distance)
        { // Verifica o critério de parada
            for (int j = 0; j < pop_size; j++)
            {
                population[j] = new_population[j];
            }
            return true;
        }
    }
    for (int i = 0; i < pop_size; i++)
    {
        population[i] = new_population[i];
    }
    return false;
}

// Encontra o melhor indivíduo na população
Individual find_best_individual(Individual population[])
{
    Individual best_individual = population[0];
    for (int i = 1; i < pop_size; i++)
    {
        if (population[i].fitness < best_individual.fitness)
        {
            best_individual = population[i];
        }
    }
    return best_individual;
}

// Função para ler coordenadas de um arquivo
void ler_coordenadas(const char *nome_arquivo, Point points[])
{
    FILE *arquivo = fopen(nome_arquivo, "r");
    if (arquivo == NULL)
    {
        printf("Erro ao abrir o arquivo.\n");
        exit(1);
    }
    int id;
    double x, y, z;
    for (int i = 0; i < n; i++)
    {
        fscanf(arquivo, "%lf %lf %lf", &x, &y, &z);
        points[i].id = i;
        points[i].x = x;
        points[i].y = y;
        points[i].z = z;
    }
    for (int i = 0; i < n; i++)
    {
        printf("ID: %d, X: %.6f, Y: %.6f, Z: %.6f\n", points[i].id, points[i].x, points[i].y, points[i].z);
    }
    fclose(arquivo);
}

// Função para calcular e imprimir a taxa de convergência
double taxa_convergencai(double best_fitness[])
{
    return (best_fitness[iterations - 1] - best_fitness[0]) / best_fitness[0];
}

double ga()
{
    // Carregar coordenadas do arquivo
    const char *nome_arquivo = "..\\coordenadas\\star100.xyz.txt";
    Point points[n];
    ler_coordenadas(nome_arquivo, points);

    // Inicializar a população
    Individual population[pop_size];
    initialize_population(population, points);

    // Array para armazenar o fitness do melhor indivíduo em cada geração
    double best_fitness[iterations];

    // Loop pricipal
    for (int i = 0; i < iterations; i++)
    {
        if (evolve_population(population, points))
        {
            printf("Parada atingida: caminho menor que %.2f encontrado\n", target_distance);
            break;
        }

        // Encontrar o melhor indivíduo da geração atual
        Individual best_individual = find_best_individual(population);
        best_fitness[i] = best_individual.fitness;

        if (i % 1000 == 0)
        {

            printf("Iteracao: %d\n", i);
            printf("Comprimento do melhor caminho: %.2f\n", best_individual.fitness);
            printf("Caminho:\n");
            for (int j = 0; j < n; j++)
            {
                printf("%d ", best_individual.path[j]);
            }
            printf("\n");
        }
    }

    // Encontrar o melhor indivíduo
    Individual best_individual = find_best_individual(population);

    // Imprimir o melhor caminho encontrado
    printf("Melhor caminho encontrado:\n");
    for (int i = 0; i < n; i++)
    {
        printf("%d ", best_individual.path[i]);
    }
    printf("\n");

    // Imprimir o comprimento do melhor caminho encontrado
    printf("Comprimento do melhor caminho: %.2f\n", best_individual.fitness);

    // Imprimir a taxa de convergência
    return taxa_convergencai(best_fitness);
}

void gravarLog(char msg[])
{
    FILE *file = fopen("log_ga.txt", "a"); // "a" para abrir em modo append (adicionar no final)
    if (file == NULL)
    {
        perror("Erro ao abrir o arquivo de log");
        return;
    }

    fprintf(file, msg);
    fclose(file);
}

void gerarGrid(int popSizeMin, int popSizeMax, float mutationRateMin, float mutationRateMax)
{
    int best_pop;
    double best_mutation;
    double best_taxa = 100;

    for (int i = popSizeMin; i < popSizeMax; i += 10)
    {
        pop_size = i;
        for (float j = mutationRateMin; j < mutationRateMax; j += 0.01)
        {
            mutation_rate = j;
            double taxa = ga();
            char msg[300];
            sprintf(msg, "[popSize:%d, mutationRate:%f] taxa de convergencia = %.6f\n", i, j, taxa);

            gravarLog(msg);

            if (taxa < best_taxa)
            {
                best_taxa = taxa;
                best_pop = i;
                best_mutation = j;
            }
        }
    }

    char msg_best[200];
    sprintf(msg_best, "Melhor combinacao: pop_size = %d , mutation_rate = %f  \nconvergencia = %.6f", best_pop, best_mutation, best_taxa);
    gravarLog(msg_best);
}

int main()
{
    // gerarGrid(10, 150, 0.01, 0.10);

    ga();
    return 0;
}

File: scripts_py/ga.py
import numpy as np
import random

N = 100                # número de pontos 3D
POP_SIZE = 10         # tamanho da população
MUTATION_RATE = 0.10   # taxa de mutação
ITERATIONS = 1000000   # número de iterações
TARGET_DISTANCE = 2000 # valor específico para o critério de parada

# Estrutura para representar um ponto 3D
class Point:
    def __init__(self, id, x, y, z):
        self.id = id
        self.x = x
        self.y = y
        self.z = z

# Estrutura para representar um indivíduo (caminho)
class Individual:
    def __init__(self, path):
        self.path = path
        self.fitness = 0.0

# Calcula a distância euclidiana entre dois pontos 3D
def distance(p1, p2):
    return np.sqrt((p2.x - p1.x)**2 + (p2.y - p1.y)**2 + (p2.z - p1.z)**2)

# Calcula o comprimento do caminho de um indivíduo
def calculate_fitness(individual, points):
    length = 0.0
    for i in range(N - 1):
        length += distance(points[individual.path[i]], points[individual.path[i + 1]])
    length += distance(points[individual.path[N - 1]], points[individual.path[0]]) # fecha o ciclo
    return length

# Inicializa uma população de indivíduos com caminhos aleatórios
def initialize_population():
    population = []
    for _ in range(POP_SIZE):
        path = list(range(N))
        random.shuffle(path)
        population.append(Individual(path))
    return population

# Seleciona indivíduos para reprodução usando torneio binário
def tournament_selection(population):
    parent1 = random.choice(population)
    parent2 = random.choice(population)
    return parent1 if parent1.fitness < parent2.fitness else parent2

# Realiza o cruzamento entre dois indivíduos para gerar um novo indivíduo
def crossover(parent1, parent2):
    start = random.randint(0, N - 1)
    end = random.randint(0, N - 1)
    if start > end:
        start, end = end, start
    child = [-1] * N
    for i in range(start, end + 1):
        child[i] = parent1.path[i]
    index = 0
    for i in range(N):
        if index == start:
            index = end + 1
        if parent2.path[i] not in child:
            child[index] = parent2.path[i]
            index += 1
    return Individual(child)

# Realiza mutação em um indivíduo
def mutate(individual):
    if random.random() < MUTATION_RATE:
        index1 = random.randint(0, N - 1)
        index2 = random.randint(0, N - 1)
        individual.path[index1], individual.path[index2] = individual.path[index2], individual.path[index1]

# Evolui a população por uma geração e verifica o critério de parada
def evolve_population(population, points):
    new_population = []
    for _ in range(POP_SIZE):
        parent1 = tournament_selection(population)
        parent2 = tournament_selection(population)
        child = crossover(parent1, parent2)
        mutate(child)
        child.fitness = calculate_fitness(child, points)
        new_population.append(child)
        if child.fitness < TARGET_DISTANCE: # Verifica o critério de parada
            return True
    population[:] = new_population
    return False

# Encontra o melhor indivíduo na população
def find_best_individual(population):
    return min(population, key=lambda x: x.fitness)

# Função para ler coordenadas de um arquivo
def ler_coordenadas(nome_arquivo):
    with open(nome_arquivo, 'r') as arquivo:
        coordenadas = []
        for linha in arquivo:
            partes = linha.strip().split()
            x, y, z = map(float, partes[:3])  # Ajuste para pegar apenas os três primeiros elementos da linha
            coordenadas.append(Point(None, x, y, z))  # Não usaremos o ID
    return coordenadas

random.seed()

# Carregar coordenadas do arquivo
nome_arquivo = "C:\\Users\\Getaruck\\Documents\\TCC\\coordenadas\\star100.xyz.txt"
points = ler_coordenadas(nome_arquivo)

# Inicializar a população
population = initialize_population()

# No loop principal, ajuste o controle de parada
for i in range(ITERATIONS):
    if evolve_population(population, points):
        print("Parada atingida: caminho menor que %.2f encontrado" % TARGET_DISTANCE)
        break
    if i % 1000 == 0:
        print("Iteração:", i)
        best_individual = find_best_individual(population)
        print("Comprimento do melhor caminho: %.2f" % best_individual.fitness)

# Encontrar o melhor indivíduo
best_individual = find_best_individual(population)

# Imprimir o melhor caminho encontrado
print("Melhor caminho encontrado:")
print(best_individual.path)

# Imprimir o comprimento do melhor caminho encontrado
print("Comprimento do melhor caminho: %.2f" % best_individual.fitness)
